### DISCUSSION TITLE
for scoping behavior

### DISCUSSION THREAD

**Discussion by castano** [shader-slang/slang#5824] [General]
When compiling hlsl shaders with slang this code:

```
for (int i = 0; i < 4; i++) {...}
for (int i = 0; i < 4; i++) {...}
```

produces the following error:

```
error 30200: declaration of 'i' conflicts with existing declaration
```

The same code is OK when the extension of the file being compiled is `.slang` instead of `.hlsl` so I imagine there's some option to toggle that behavior. Is it possible to change this independently of the file extension?

While in HLSL prior to version 2021, for variables were not scoped to the body of the for loop, that code was legal, because the second declaration would simply shadow the first one. So, slang's behavior is not particularly helpful.

Since HLSL 2021 for variable declarations are scoped differently. It would be very useful to be able to replicate the new behavior, which also appears to be Slang's default. See: https://devblogs.microsoft.com/directx/hlsl-2021-migration-guide/

**Comment by jkwak-work** [comment]
I am adding a new section on our document to explain this behavior.
https://github.com/shader-slang/shader-slang.github.io/pull/32

In short, you can set the input language to be "slang" not "hlsl".
If the file extension is ".slang", the file will be automatically treated as Slang shader.
But you can use ".hlsl" file extension and set it explicitly to be Slang shader to avoid the problem.

**Comment by csyonghe** [comment]
We recommend our users to not specify HLSL as the input language even if the source is in HLSL. It is not easy to define the exact behavior of HLSL that Slang should follow, and the current implementation of -lang HLSL will turn on some legacy HLSL behavior that not all users would want. In 99% of the case skipping the -lang setting should work just fine without introducing the true legacy HLSL behaviors.

**Comment by csyonghe** [comment]
When you are compiling HLSL files through slangc, we recommend setting -lang slang in the command line to opt out of legacy HLSL behavior.
